ang/civicase-base/decorators/crm-api.decorator.js

```diff
@@ -0,0 +1,35 @@
+(function (angular, $, _) {
+  var module = angular.module('civicase-base');
+
+  /**
+   * The original crmApi, which is part of `crmUtil` module, does not 'reject'
+   * the promise when an array of api requests are sent, and one of them got
+   * returned with error message. This decorator fixes that.
+   */
+  module.decorator('crmApi', function ($delegate, $q) {
+    var crmApi = function (entity, action, params, message) {
+      var deferred = $q.defer();
+
+      $delegate(entity, action, params, message)
+        .then(function (result) {
+          var isError = false;
+
+          if (_.isArray(result)) {
+            _.each(result, function (response) {
+              if (response.is_error) {
+                isError = true;
+              }
+            });
+          }
+
+          isError ? deferred.reject(result) : deferred.resolve(result);
+        });
+
+      return deferred.promise;
+    };
+
+    crmApi.val = $delegate.val;
+
+    return crmApi;
+  });
+})(angular, CRM.$, CRM._);
```

Comment:
Decorating the CRM Api is a good idea, but we are affecting all calls. What happens if a specific call was taking into consideration that promises are not rejected even if they have an error? Did we check this? I'm asking because there's a lot of core code that might have relied on this "feature".

Have we also considered fixing core? Might take longer, but it would be better in the long run to avoid regressions in the future.

Reply:
This is a very good point and its true. In core there are many such places. Lets have a discussion about this.